File: ein.py/src/ein_bot/trace/render.py
"""Markdown trace renderer — S1.6.4 T1.6.4.2 / .3.

Threads a linearized :class:`~ein_bot.trace.linearize.Trace` into a
single self-contained markdown narrative: a numbered step per rule
firing (name, English ``:why``, premises with their quoted source
sentences, and an inline `dot` derivation slice), refuted hypotheses
folded into `<details>` reductio sections, and a closing lattice-DAG +
solution grid. Every diagram is an inline fenced `dot` block — no SVG
(the Python layer emits DOT; rasterising is a shell concern).

``mode="reorder"`` clusters the steps by the entity they are about
(`## About <X>`) — a presentation pass over the same steps (T1.6.4.3).
"""
from __future__ import annotations

from .ast import TraceStep
from .linearize import Reductio, Trace


def _dot_block(dot: str | None) -> list[str]:
    if not dot:
        return []
    return ["```dot", dot, "```", ""]


def _premises_line(step: TraceStep) -> str:
    labels = step.premise_labels()
    if labels:
        base = "Premises: " + ", ".join(f"`{p}`" for p in labels)
        if step.sources:                       # source-sentence quoting (T1.6.4.5)
            base += " — from " + ", ".join(step.sources)
        return base
    if step.sources:                           # no derived premise — the source is given
        return "Premises: from " + ", ".join(step.sources)
    return "Premises: —"


def _render_step(step: TraceStep, *, diagrams: bool) -> list[str]:
    out = [f"## Step {step.n} — `{step.rule}`", ""]
    if step.why:
        out += [f"> {step.why}", ""]
    out += [_premises_line(step), "",
            f"Derives `{step.derived_label}`.", ""]
    if diagrams:
        out += _dot_block(step.diagram)
    return out


def _render_reductio(r: Reductio, *, diagrams: bool) -> list[str]:
    out = ["<details>", f"<summary>{r.summary}</summary>", ""]
    out += [f"Assumed **{r.commitment}**; the branch derives ⊥.", ""]
    if r.learned_clause:
        out += [f"Lifted no-good: `{r.learned_clause}`.", ""]
    if diagrams:
        out += _dot_block(r.diagram)
    out += ["</details>", ""]
    return out


def render_markdown(
    trace: Trace, *, mode: str = "engine", diagrams: bool = True,
) -> str:
    """Render a :class:`Trace` as a markdown string.

    ``mode`` is ``"engine"`` (numbered engine order, default) or
    ``"reorder"`` (clustered by target entity).
    """
    if mode not in ("engine", "reorder"):
        raise ValueError(f"unknown trace mode: {mode!r} (expected 'engine' or 'reorder')")

    lines = ["# Solution trace", "", f"> {trace.summary}", ""]
    if trace.commitment and trace.commitment not in ("∅ (unconditional)", "—"):
        lines += [f"Assuming **{trace.commitment}**.", ""]

    if not trace.steps:
        lines += ["_(no surviving derivation — see the refuted branches below.)_", ""]
    elif mode == "reorder":
        lines += _render_reordered(trace, diagrams=diagrams)
    else:
        for step in trace.steps:
            lines += _render_step(step, diagrams=diagrams)

    if trace.reductios:
        lines += ["## Refuted hypotheses", ""]
        for r in trace.reductios:
            lines += _render_reductio(r, diagrams=diagrams)

    if diagrams and trace.lattice_dot:
        lines += ["## Commitment lattice", "", *_dot_block(trace.lattice_dot)]
    if diagrams and trace.solution_dot:
        lines += ["## Solution", "", *_dot_block(trace.solution_dot)]
    if trace.full_kb_dot:
        lines += ["## Full KB (final state)", "", *_dot_block(trace.full_kb_dot)]

    return "\n".join(lines).rstrip() + "\n"


def _render_reordered(trace: Trace, *, diagrams: bool) -> list[str]:
    """Stable-partition the steps by their target entity (T1.6.4.3).

    A presentation pass: every step is emitted exactly once, keeping its
    engine step number and the within-cluster order — so the reordered
    trace has the *same set of steps* as engine order (the acceptance
    criterion), grouped under `## About <entity>` headings. Clusters
    appear in first-seen order; steps with no entity fall under "About
    (misc)".
    """
    order: list[str] = []
    by_entity: dict[str, list[TraceStep]] = {}
    for step in trace.steps:
        key = step.section or "(misc)"
        if key not in by_entity:
            by_entity[key] = []
            order.append(key)
        by_entity[key].append(step)

    out: list[str] = []
    for entity in order:
        out += [f"## About {entity}", ""]
        for step in by_entity[entity]:
            why = f" — {step.why}" if step.why else ""
            out += [f"**Step {step.n}** · `{step.rule}` → `{step.derived_label}`{why}", ""]
            out.append(_premises_line(step))
            out.append("")
            if diagrams:
                out += _dot_block(step.diagram)
    return out


__all__ = ["render_markdown"]

File: ein.rs/crates/ein-render/src/trace/render.rs
//! The markdown trace renderer — `ein.py`'s `trace/render.py`.
//!
//! Threads a linearised [`Trace`] into one self-contained markdown narrative:
//! a numbered step per rule firing (name, English `:why`, premises with their
//! quoted source sentences, and an inline `dot` derivation slice), refuted
//! hypotheses folded into `<details>` reductio sections, and a closing
//! lattice DAG plus solution grid. Every diagram is an inline fenced `dot`
//! block — no SVG; the engine emits DOT and rasterising is a shell concern.
//!
//! `Mode::Reorder` clusters the steps by the entity they are about — a
//! presentation pass over the same steps, emitting each exactly once.

use super::ast::TraceStep;
use super::linearize::{Reductio, Trace};

/// Numbered engine order (the default), or clustered by target entity.
#[derive(Clone, Copy, PartialEq, Eq)]
pub enum Mode {
    Engine,
    Reorder,
}

impl Mode {
    pub fn parse(s: &str) -> Option<Mode> {
        match s {
            "engine" => Some(Mode::Engine),
            "reorder" => Some(Mode::Reorder),
            _ => None,
        }
    }
}

fn dot_block(dot: Option<&String>) -> Vec<String> {
    match dot.filter(|d| !d.is_empty()) {
        None => Vec::new(),
        Some(d) => vec![
            "```dot".to_string(),
            d.clone(),
            "```".to_string(),
            String::new(),
        ],
    }
}

fn premises_line(step: &TraceStep) -> String {
    let labels = step.premise_labels();
    if !labels.is_empty() {
        let quoted: Vec<String> = labels.iter().map(|p| format!("`{p}`")).collect();
        let mut base = format!("Premises: {}", quoted.join(", "));
        if !step.sources.is_empty() {
            base.push_str(&format!(" — from {}", step.sources.join(", ")));
        }
        return base;
    }
    if !step.sources.is_empty() {
        // No derived premise — the source *is* what was given.
        return format!("Premises: from {}", step.sources.join(", "));
    }
    "Premises: —".to_string()
}

fn render_step(step: &TraceStep, diagrams: bool) -> Vec<String> {
    let mut out = vec![
        format!("## Step {} — `{}`", step.n, step.rule),
        String::new(),
    ];
    if !step.why.is_empty() {
        out.push(format!("> {}", step.why));
        out.push(String::new());
    }
    out.push(premises_line(step));
    out.push(String::new());
    out.push(format!("Derives `{}`.", step.derived_label()));
    out.push(String::new());
    if diagrams {
        out.extend(dot_block(step.diagram.as_ref()));
    }
    out
}

fn render_reductio(r: &Reductio, diagrams: bool) -> Vec<String> {
    let mut out = vec![
        "<details>".to_string(),
        format!("<summary>{}</summary>", r.summary),
        String::new(),
        format!("Assumed **{}**; the branch derives ⊥.", r.commitment),
        String::new(),
    ];
    if !r.learned_clause.is_empty() {
        out.push(format!("Lifted no-good: `{}`.", r.learned_clause));
        out.push(String::new());
    }
    if diagrams {
        out.extend(dot_block(r.diagram.as_ref()));
    }
    out.push("</details>".to_string());
    out.push(String::new());
    out
}

/// Render a [`Trace`] as markdown.
pub fn render_markdown(trace: &Trace, mode: Mode, diagrams: bool) -> String {
    let mut lines = vec![
        "# Solution trace".to_string(),
        String::new(),
        format!("> {}", trace.summary),
        String::new(),
    ];
    if !trace.commitment.is_empty()
        && trace.commitment != "∅ (unconditional)"
        && trace.commitment != "—"
    {
        lines.push(format!("Assuming **{}**.", trace.commitment));
        lines.push(String::new());
    }

    if trace.steps.is_empty() {
        lines.push("_(no surviving derivation — see the refuted branches below.)_".to_string());
        lines.push(String::new());
    } else if mode == Mode::Reorder {
        lines.extend(render_reordered(trace, diagrams));
    } else {
        let mut emitted_hyp = false;
        for step in &trace.steps {
            // The unconditional spine first, then the under-hypothesis block.
            if step.conditional && !emitted_hyp {
                lines.push(format!("## Under hypothesis — {}", trace.commitment));
                lines.push(String::new());
                emitted_hyp = true;
            }
            lines.extend(render_step(step, diagrams));
        }
    }

    if !trace.reductios.is_empty() {
        lines.push("## Refuted hypotheses".to_string());
        lines.push(String::new());
        for r in &trace.reductios {
            lines.extend(render_reductio(r, diagrams));
        }
    }

    if diagrams && trace.lattice_dot.is_some() {
        lines.push("## Commitment lattice".to_string());
        lines.push(String::new());
        lines.extend(dot_block(trace.lattice_dot.as_ref()));
    }
    if diagrams && trace.solution_dot.is_some() {
        lines.push("## Solution".to_string());
        lines.push(String::new());
        lines.extend(dot_block(trace.solution_dot.as_ref()));
    }
    if trace.full_kb_dot.is_some() {
        lines.push("## Full KB (final state)".to_string());
        lines.push(String::new());
        lines.extend(dot_block(trace.full_kb_dot.as_ref()));
    }

    format!("{}\n", lines.join("\n").trim_end())
}

/// Stable-partition the steps by their target entity.
///
/// A presentation pass: every step is emitted exactly once, keeping its engine
/// step number and its within-cluster order — so the reordered trace has the
/// *same set of steps* as engine order, grouped under `## About <entity>`.
/// Clusters appear in first-seen order; a step with no entity falls under
/// "About (misc)".
fn render_reordered(trace: &Trace, diagrams: bool) -> Vec<String> {
    let mut order: Vec<String> = Vec::new();
    let mut by_entity: Vec<(String, Vec<&TraceStep>)> = Vec::new();
    for step in &trace.steps {
        let key = step.section.clone().unwrap_or_else(|| "(misc)".to_string());
        match by_entity.iter_mut().find(|(k, _)| *k == key) {
            Some(slot) => slot.1.push(step),
            None => {
                order.push(key.clone());
                by_entity.push((key, vec![step]));
            }
        }
    }
    let mut out: Vec<String> = Vec::new();
    for entity in &order {
        out.push(format!("## About {entity}"));
        out.push(String::new());
        let steps = by_entity
            .iter()
            .find(|(k, _)| k == entity)
            .map(|(_, v)| v)
            .expect("registered above");
        for step in steps {
            let why = if step.why.is_empty() {
                String::new()
            } else {
                format!(" — {}", step.why)
            };
            out.push(format!(
                "**Step {}** · `{}` → `{}`{why}",
                step.n,
                step.rule,
                step.derived_label()
            ));
            out.push(String::new());
            out.push(premises_line(step));
            out.push(String::new());
            if diagrams {
                out.extend(dot_block(step.diagram.as_ref()));
            }
        }
    }
    out
}
